### PR TITLE
types: narrow code reply type based on schema

### PR DESCRIPTION
**Pull request by hpx7**
Closes https://github.com/fastify/fastify/issues/5378

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/reply.d.ts [hunk 1]
```diff
@@ -23,7 +23,9 @@ type ReplyTypeConstrainer<RouteGenericReply, Code extends ReplyKeysToCodes<keyof
 export type ResolveReplyTypeWithRouteGeneric<RouteGenericReply, Code extends ReplyKeysToCodes<keyof RouteGenericReply>,
   SchemaCompiler extends FastifySchema = FastifySchema,
   TypeProvider extends FastifyTypeProvider = FastifyTypeProviderDefault> =
-  ResolveFastifyReplyType<TypeProvider, SchemaCompiler, { Reply: ReplyTypeConstrainer<RouteGenericReply, Code> }>
+  Code extends keyof SchemaCompiler['response'] ?
```
Comment (climba03003) [inline types/reply.d.ts:26]: What if the status code is `2xx` or `default`?

**Review by climba03003 — CHANGES_REQUESTED** [review]

**Comment by climba03003** [comment]
For others who would like to review this PR. We have two direction here

1. Take a small step first and accept the incomplete of `status code` type-safe.
That means `2xx` or `default` remain no type-safe and exact `status code` would beneficial.

2. Take a complete step and not accept a partial solution.

File: test/types/type-provider.test-d.ts [hunk 2]
```diff
@@ -469,7 +469,7 @@ expectAssignable(server.withTypeProvider<TypeBoxProvider>().get(
     res.send('hello')
     res.send(42)
     res.send({ error: 'error' })
-    expectType<(payload?: string | number | { error: string }) => typeof res>(res.code(200).send)
+    expectType<(payload?: string) => typeof res>(res.code(200).send)
```
Comment (mcollina) [inline test/types/type-provider.test-d.ts:472]: Can you please add more expectType assertions for the other status codes?

**Review by mcollina — COMMENTED** [review]
I think this is an improvement worth doing.

File: test/types/type-provider.test-d.ts (hunk 2 above)
Comment (hpx7) [reply test/types/type-provider.test-d.ts:472]: done 👍

**Comment by climba03003** [comment]
I have updated the branch by rebase.

**Review by climba03003 — APPROVED** [review]

**Comment by fox1t** [comment]
I agree that at least fixing the 1. should be a tangible improvement anyway.

**Review by fox1t — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**